app/code/Magento/EncryptionKey/Setup/Patch/Data/SodiumChachaPatch.php

```diff
@@ -0,0 +1,129 @@
+
+    /**
+     * @var \Magento\Framework\Encryption\EncryptorInterface
+     */
+    private $encryptor;
+
+    /**
+     * @var \Magento\Framework\Config\ScopeInterface
+     */
+    private $scope;
+
+    /**
+     * SodiumChachaPatch constructor.
+     * @param \Magento\Framework\Setup\ModuleDataSetupInterface $moduleDataSetup
+     * @param \Magento\Config\Model\Config\Structure\Proxy $structure
+     * @param \Magento\Framework\Encryption\EncryptorInterface $encryptor
+     * @param \Magento\Framework\Config\ScopeInterface $scope
+     */
+    public function __construct(
+        \Magento\Framework\Setup\ModuleDataSetupInterface $moduleDataSetup,
+        \Magento\Config\Model\Config\Structure\Proxy $structure,
+        \Magento\Framework\Encryption\EncryptorInterface $encryptor,
+        \Magento\Framework\Config\ScopeInterface $scope
+    ) {
+        $this->moduleDataSetup = $moduleDataSetup;
+        $this->structure = $structure;
+        $this->encryptor = $encryptor;
+        $this->scope = $scope;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply()
+    {
+        $this->moduleDataSetup->startSetup();
+
+        $this->reEncryptSystemConfigurationValues();
+        $this->reEncryptCreditCardNumbers();
+
+        $this->moduleDataSetup->endSetup();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getDependencies()
+    {
+        return [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAliases()
+    {
+        return [];
+    }
+
+    private function reEncryptSystemConfigurationValues()
+    {
+        $currentScope = $this->scope->getCurrentScope();
+
+        $this->scope->setCurrentScope(\Magento\Framework\App\Area::AREA_ADMINHTML);
+
+        $paths = $this->structure->getFieldPathsByAttribute(
+            'backend_model',
+            \Magento\Config\Model\Config\Backend\Encrypted::class
+        );
+
+        $this->scope->setCurrentScope($currentScope);
+
+        // walk through found data and re-encrypt it
+        if ($paths) {
+            $table = $this->moduleDataSetup->getTable('core_config_data');
+            $values = $this->moduleDataSetup->getConnection()->fetchPairs(
+                $this->moduleDataSetup->getConnection()
+                    ->select()
+                    ->from($table, ['config_id', 'value'])
+                    ->where('path IN (?)', $paths)
+                    ->where('value NOT LIKE ?', '')
+            );
+            foreach ($values as $configId => $value) {
+                $this->moduleDataSetup->getConnection()->update(
+                    $table,
+                    ['value' => $this->encryptor->encrypt($this->encryptor->decrypt($value))],
+                    ['config_id = ?' => (int)$configId]
+                );
+            }
+        }
+    }
+
+    private function reEncryptCreditCardNumbers()
+    {
+        $table = $this->moduleDataSetup->getTable('sales_order_payment');
```

Comment:
This data can be large.
I'd just leave the data encrypted, so it will be re-encrypted when read/written next time.
Additionally we could provide a separate tool or Magento CLI command to convert encrypted data. This can be run separately on already upgraded Magento instance. Both old and new data is usable for Magento application, so users can run the tool only if they really want to migrate to the newer algorithm (which should not be a common case now, as current algorithm still pretty good as I understand).